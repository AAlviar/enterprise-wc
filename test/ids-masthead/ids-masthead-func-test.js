/**
 * @jest-environment jsdom
 */
import '../helpers/resize-observer-mock';
import '../helpers/match-media-mock';
import IdsMasthead from '../../src/components/ids-masthead/ids-masthead';

describe('IdsMasthead Component', () => {
  let element;
  let sections;

  beforeEach(async () => {
    const masthead = new IdsMasthead();
    masthead.icon = 'logo';
    masthead.title = 'title';
    masthead.innerHTML = `
      <nav slot="start"><ids-button>start nav</ids-button></nav>
      <nav slot="center"><ids-menu-button>center nav</ids-button></nav>
      <nav slot="end"><ids-button>end nav</ids-button></nav>
    `;

    document.body.appendChild(masthead);
    element = document.querySelector('ids-masthead');
    sections = element.elements.sections;
  });

  afterEach(async () => {
    document.body.innerHTML = '';
  });

  it('renders with no errors', () => {
    const errors = jest.spyOn(global.console, 'error');
    expect(document.querySelectorAll('ids-masthead').length).toEqual(1);
    expect(errors).not.toHaveBeenCalled();

    element.remove();
    element = new IdsMasthead();
    document.body.appendChild(element);
    expect(document.querySelectorAll('ids-masthead').length).toEqual(1);
    expect(errors).not.toHaveBeenCalled();
  });

  it('renders correctly', () => {
    element.logo = '';
    element.title = '';
    expect(element.outerHTML).toMatchSnapshot();

    element.icon = 'info';
    expect(element.outerHTML).toMatchSnapshot();

    element.icon = 'logo';
    element.title = 'Infor Application';
    expect(element.outerHTML).toMatchSnapshot();
  });

  it('renders any icon', () => {
    element.setAttribute('icon', '');
    expect(element.icon).toBe('');

    element.icon = 'info';
    expect(element.icon).toBe('info');
    expect(element.getAttribute('icon')).toBe('info');
    expect(element.template()).toMatchSnapshot();

    element.icon = 'user';
    expect(element.icon).toBe('user');
    expect(element.getAttribute('icon')).toBe('user');
    expect(element.template()).toMatchSnapshot();
  });

  it('renders logo', () => {
    element.setAttribute('icon', '');
    expect(element.icon).toBe('');

    element.icon = 'logo';
    expect(element.icon).toBe('logo');
    expect(element.getAttribute('icon')).toBe('logo');
    expect(element.template()).toMatchSnapshot();
  });

  it('renders clickable logo', () => {});

  it('renders title', () => {
    element.setAttribute('title', '');
    expect(element.title).toBe('');
    element.title = 'Info Application';
    expect(element.title).toBe('Info Application');
    expect(element.getAttribute('title')).toBe('Info Application');
    expect(element.template()).toMatchSnapshot();
  });

  it('restyles buttons to be square and transparent', () => {
    const buttons = element.querySelectorAll('ids-button, ids-menu-button');
    buttons.forEach((button) => {
      expect(button.colorVariant).toBe('alternate');
      expect(button.square).toBe(true);
      expect(button.type).toBe('default');
    });
  });

  it('has breakpoints', () => {
    expect(element.breakpoints.mobile).toBeDefined();
    expect(element.breakpoints.tablet).toBeDefined();
    expect(element.breakpoints.desktop).toBeDefined();

    expect(element.breakpoints.mobile.addEventListener).toHaveBeenCalled();
    expect(element.breakpoints.tablet.addEventListener).toHaveBeenCalled();
    expect(element.breakpoints.desktop.addEventListener).toHaveBeenCalled();

    expect(global.matchMedia).toHaveBeenCalled();
  });

  it('renders breakpoint: desktop', () => {
    jest.spyOn(element, 'isDesktop', 'get').mockReturnValue(true);
    expect(element.isDesktop).toBe(true);
    expect(element.isTablet).toBe(false);
    expect(element.isMobile).toBe(false);

    element.renderBreakpoint();
    expect(sections.start.querySelector('slot').assignedNodes()).toContain(element.slots.start);
    expect(sections.center.querySelector('slot').assignedNodes()).toContain(element.slots.center);
    expect(sections.end.querySelector('slot').assignedNodes()).toContain(element.slots.end);

    expect(sections.more.querySelector('slot').assignedNodes()).toEqual([]);
  });

  it('renders breakpoint: tablet', () => {
    jest.spyOn(element, 'isTablet', 'get').mockReturnValue(true);
    expect(element.isTablet).toBe(true);
    expect(element.isDesktop).toBe(false);
    expect(element.isMobile).toBe(false);

    element.renderBreakpoint();
    expect(sections.start.querySelector('slot').assignedNodes()).toContain(element.slots.start);
    expect(sections.center.querySelector('slot').assignedNodes()).toEqual([]);
    expect(sections.end.querySelector('slot').assignedNodes()).toEqual([]);

    expect(sections.more.querySelector('slot').assignedNodes()).toContain(element.slots.center);
    expect(sections.more.querySelector('slot').assignedNodes()).toContain(element.slots.end);
  });

  it('renders breakpoint: mobile', () => {
    jest.spyOn(element, 'isMobile', 'get').mockReturnValue(true);
    expect(element.isMobile).toBe(true);
    expect(element.isDesktop).toBe(false);
    expect(element.isTablet).toBe(false);

    element.renderBreakpoint();
    expect(sections.start.querySelector('slot').assignedNodes()).toEqual([]);
    expect(sections.center.querySelector('slot').assignedNodes()).toEqual([]);
    expect(sections.end.querySelector('slot').assignedNodes()).toEqual([]);

    expect(sections.more.querySelector('slot').assignedNodes()).toContain(element.slots.start);
    expect(sections.more.querySelector('slot').assignedNodes()).toContain(element.slots.center);
    expect(sections.more.querySelector('slot').assignedNodes()).toContain(element.slots.end);
  });
});
