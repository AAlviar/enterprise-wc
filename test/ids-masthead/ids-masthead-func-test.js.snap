// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IdsMasthead Component renders any icon 1`] = `
"
      <ids-toolbar class=\\"ids-masthead\\" tabbable=\\"true\\">
        <ids-toolbar-section id=\\"start\\" align=\\"start\\" type=\\"fluid\\">
          
        <ids-button id=\\"logo\\" class=\\"icon-info\\" color-variant=\\"alternate\\" square=\\"true\\">
          <ids-icon slot=\\"icon\\" icon=\\"info\\"></ids-icon>
          <ids-text slot=\\"text\\" audible=\\"true\\">Masthead logo</ids-text>
        </ids-button>
       <ids-text id=\\"title\\" color-variant=\\"alternate\\" font-size=\\"14\\" font-weight=\\"bold\\">title</ids-text>
          <slot name=\\"start\\"></slot>
        </ids-toolbar-section>

        <ids-toolbar-section id=\\"center\\" align=\\"center\\" type=\\"fluid\\">
          <slot name=\\"center\\"></slot>
        </ids-toolbar-section>

        <ids-toolbar-section id=\\"end\\" align=\\"end\\" type=\\"fluid\\">
          <slot name=\\"end\\"></slot>
        </ids-toolbar-section>

        <ids-toolbar-more-actions id=\\"more\\" icon=\\"expand-app-tray\\" class=\\"hidden\\">
          <slot name=\\"more\\"></slot>
        </ids-toolbar-more-actions>
      </ids-toolbar>
    "
`;

exports[`IdsMasthead Component renders any icon 2`] = `
"
      <ids-toolbar class=\\"ids-masthead\\" tabbable=\\"true\\">
        <ids-toolbar-section id=\\"start\\" align=\\"start\\" type=\\"fluid\\">
          
        <ids-button id=\\"logo\\" class=\\"icon-user\\" color-variant=\\"alternate\\" square=\\"true\\">
          <ids-icon slot=\\"icon\\" icon=\\"user\\"></ids-icon>
          <ids-text slot=\\"text\\" audible=\\"true\\">Masthead logo</ids-text>
        </ids-button>
       <ids-text id=\\"title\\" color-variant=\\"alternate\\" font-size=\\"14\\" font-weight=\\"bold\\">title</ids-text>
          <slot name=\\"start\\"></slot>
        </ids-toolbar-section>

        <ids-toolbar-section id=\\"center\\" align=\\"center\\" type=\\"fluid\\">
          <slot name=\\"center\\"></slot>
        </ids-toolbar-section>

        <ids-toolbar-section id=\\"end\\" align=\\"end\\" type=\\"fluid\\">
          <slot name=\\"end\\"></slot>
        </ids-toolbar-section>

        <ids-toolbar-more-actions id=\\"more\\" icon=\\"expand-app-tray\\" class=\\"hidden\\">
          <slot name=\\"more\\"></slot>
        </ids-toolbar-more-actions>
      </ids-toolbar>
    "
`;

exports[`IdsMasthead Component renders correctly 1`] = `
"<ids-masthead icon=\\"logo\\" title=\\"\\">
      <nav slot=\\"start\\"><ids-button color-variant=\\"alternate\\" language=\\"en\\" locale=\\"en-US\\" square=\\"\\">start nav</ids-button></nav>
      <nav slot=\\"center\\"><ids-menu-button color-variant=\\"alternate\\" language=\\"en\\" locale=\\"en-US\\" square=\\"\\">center nav</ids-menu-button></nav>
      <nav slot=\\"end\\"><ids-button color-variant=\\"alternate\\" language=\\"en\\" locale=\\"en-US\\" square=\\"\\">end nav</ids-button></nav>
    </ids-masthead>"
`;

exports[`IdsMasthead Component renders correctly 2`] = `
"<ids-masthead icon=\\"info\\" title=\\"\\">
      <nav slot=\\"start\\"><ids-button color-variant=\\"alternate\\" language=\\"en\\" locale=\\"en-US\\" square=\\"\\">start nav</ids-button></nav>
      <nav slot=\\"center\\"><ids-menu-button color-variant=\\"alternate\\" language=\\"en\\" locale=\\"en-US\\" square=\\"\\">center nav</ids-menu-button></nav>
      <nav slot=\\"end\\"><ids-button color-variant=\\"alternate\\" language=\\"en\\" locale=\\"en-US\\" square=\\"\\">end nav</ids-button></nav>
    </ids-masthead>"
`;

exports[`IdsMasthead Component renders correctly 3`] = `
"<ids-masthead icon=\\"logo\\" title=\\"Infor Application\\">
      <nav slot=\\"start\\"><ids-button color-variant=\\"alternate\\" language=\\"en\\" locale=\\"en-US\\" square=\\"\\">start nav</ids-button></nav>
      <nav slot=\\"center\\"><ids-menu-button color-variant=\\"alternate\\" language=\\"en\\" locale=\\"en-US\\" square=\\"\\">center nav</ids-menu-button></nav>
      <nav slot=\\"end\\"><ids-button color-variant=\\"alternate\\" language=\\"en\\" locale=\\"en-US\\" square=\\"\\">end nav</ids-button></nav>
    </ids-masthead>"
`;

exports[`IdsMasthead Component renders logo 1`] = `
"
      <ids-toolbar class=\\"ids-masthead\\" tabbable=\\"true\\">
        <ids-toolbar-section id=\\"start\\" align=\\"start\\" type=\\"fluid\\">
          
        <ids-button id=\\"logo\\" class=\\"icon-logo\\" color-variant=\\"alternate\\" square=\\"true\\">
          <ids-icon slot=\\"icon\\" icon=\\"logo\\" viewbox=\\"0 0 32 34\\" width=\\"30\\" height=\\"30\\"></ids-icon>
          <ids-text slot=\\"text\\" audible=\\"true\\">Masthead logo</ids-text>
        </ids-button>
       <ids-text id=\\"title\\" color-variant=\\"alternate\\" font-size=\\"14\\" font-weight=\\"bold\\">title</ids-text>
          <slot name=\\"start\\"></slot>
        </ids-toolbar-section>

        <ids-toolbar-section id=\\"center\\" align=\\"center\\" type=\\"fluid\\">
          <slot name=\\"center\\"></slot>
        </ids-toolbar-section>

        <ids-toolbar-section id=\\"end\\" align=\\"end\\" type=\\"fluid\\">
          <slot name=\\"end\\"></slot>
        </ids-toolbar-section>

        <ids-toolbar-more-actions id=\\"more\\" icon=\\"expand-app-tray\\" class=\\"hidden\\">
          <slot name=\\"more\\"></slot>
        </ids-toolbar-more-actions>
      </ids-toolbar>
    "
`;

exports[`IdsMasthead Component renders title 1`] = `
"
      <ids-toolbar class=\\"ids-masthead\\" tabbable=\\"true\\">
        <ids-toolbar-section id=\\"start\\" align=\\"start\\" type=\\"fluid\\">
          
        <ids-button id=\\"logo\\" class=\\"icon-logo\\" color-variant=\\"alternate\\" square=\\"true\\">
          <ids-icon slot=\\"icon\\" icon=\\"logo\\" viewbox=\\"0 0 32 34\\" width=\\"30\\" height=\\"30\\"></ids-icon>
          <ids-text slot=\\"text\\" audible=\\"true\\">Masthead logo</ids-text>
        </ids-button>
       <ids-text id=\\"title\\" color-variant=\\"alternate\\" font-size=\\"14\\" font-weight=\\"bold\\">Info Application</ids-text>
          <slot name=\\"start\\"></slot>
        </ids-toolbar-section>

        <ids-toolbar-section id=\\"center\\" align=\\"center\\" type=\\"fluid\\">
          <slot name=\\"center\\"></slot>
        </ids-toolbar-section>

        <ids-toolbar-section id=\\"end\\" align=\\"end\\" type=\\"fluid\\">
          <slot name=\\"end\\"></slot>
        </ids-toolbar-section>

        <ids-toolbar-more-actions id=\\"more\\" icon=\\"expand-app-tray\\" class=\\"hidden\\">
          <slot name=\\"more\\"></slot>
        </ids-toolbar-more-actions>
      </ids-toolbar>
    "
`;
